<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>SARS-Cov-2 Analysis</title>
    <!-- styles -->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css" integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh" crossorigin="anonymous">
    <link rel="stylesheet" href="file://{{ css_file }}">
    <link rel="stylesheet" href="https://pro.fontawesome.com/releases/v5.13.0/css/all.css">
</head>
<body>
    <div class="header">
        <div class="frontsum row no-gutters">
            <div id="lablogo" class="col-6">
                <img class="lablogo" src="file://{{ lablogo }}" ></i>
            </div>
            <div id="patientinfo" class="col-2 infohead">
                <div class="infoheader">
                    <p>
                        PATIENT
                    </p>
                </div>
                <div class="infobody">
                    <b>NAME:</b> {{ first_name }} {{last_name}}<br>
                    <b>ACC#:</b> {{ sample_id }}<br>
                    <b>DOB:</b> {{ dob }}<br>
                    <b>SEX:</b> {{ sex }} 
                </div>
            </div>
            <div id="specimeninfo" class="col-2 infohead">
                <div class="infoheader">
                    <p>
                        SPECIMEN
                    </p>
                </div>
                <div class="infobody">
                    <b>COLLECTION:</b> {{ collection_date }}<br>
                    <b>RECEIVED:</b> {{ received_date }}<br>
                    <b>REPORT:</b> {{ report_date }}
                </div>
            </div>
            <div id="providerinfo" class="col-2 infohead">
                <div class="infoheader">
                    <p>
                        PROVIDER
                    </p>
                </div>
                <div class="infobody">
                    {{ lab_name }}<br>
                    {{ lab_street }}<br>
                    {{ lab_city }}, {{ lab_state }} {{ lab_zip }}<br>
                    {{ lab_dir }}
                </div>
            </div>
        </div>
        <div class="row titlebar" id="titlebar">
            <h4 class="headertitle">
                {{ report_title }}
            </h4>
        </div>
    </div>
    <div id="summary">
        <div class="headline">
            <h1>Summary</h1>
        </div>
        <p class="summary-text explainer">See further report pages for more detailed information about the results and quality controls.</p>
        <table class="summarytable">
            <tr>
                <td>Results</td>
                <td>{{ summary["results"] }}</td>
            </tr>
            <tr>
                <td>Lineage</td>
                <td>{{ summary["lineage"] }}</td>
            </tr>
            <tr>
                <td>Taxon</td>
                <td>{{ summary["taxon"] }}</td>
            </tr>
            <tr>
                <td>Clade</td>
                <td>{{ summary["clade"] }}</td>
            </tr>
            <tr>
                <td>Parent clade</td>
                <td>{{ summary["parent clade"] }}</td>
            </tr>
            <tr>
                <td>Confidence</td>
                <td>{{ summary["confidence"] }}</td>
            </tr>
            <tr>
                <td>Quality control</td>
                <td>{{ summary["qc"] }}</td>
            </tr>
            <tr>
                <td>Other info</td>
                <td>{{ summary["other"] }}</td>
            </tr>
            <tr>
                <th>Gene</th>
                <th>Patient info</th>
                <th>Alleles tested</th>
            </tr>
        </table>
        <p class="summary-end explainer">The SARS-CoV-2 lineage was determined by calling genetic variants of the sample compared to a reference genome (1_NC_045512.2:1).
            This is performed by the mapping of sequencing reads from the input sample to the reference genome,
            then performing a filtration of low quality alignments or variations via metrics of coverage and other factors.
            During this process, phylogeny and viral clade were also determined and presented within the report.</p>
    </div>
<div id="tree">
    <div class="headline tree">
        <h1>Phylogeny</h1>
    </div>
    <p class="treeexplain explainer">The variation present in the sample allow insights into the origins of the virus. This phylogenetic tree uses the GSAID sample database (June 14 2021) and a model for estimating maximum likelihood phylogenies.</p>
    <div class="phylocontainer">
        <img class="phylo" src="file://{{ tree_path }}">
    </div>
    <div class="legendcontainer">
        <img class="legend" src="file://{{ legend_path }}">
    </div>
<div id="prevalence">
    <div class="headline prevalencehead">
        <h1>Strain prevalence</h1>
    </div>
    <p class="prevalencexplain explainer">The following map represents the abundance of the sampleâ€™s strain among geographical regions. GenBank and GSAID data were used to create this map visualization. Strain prevalence in the map is a direct measurement of sample count in the databases.</p>
    <div class="mapcontainer">
        <img class="map" src="file://{{ map_path }}">
    </div>
</div>
<div id=variants>
    <div class="headline variantshead">
      <h1>Variant Table</h1>
    </div>
    <p class="variantsexplain explainer">These are the variations from the reference genome discovered in the sample. The reference genome used is NC_045512.2 and the position represents the genomic position of the variation, aligned to the left-most base when necessary.</p>
    <table class="variants table">
      <tr>
        <th>Position</th>
        <th>Reference</th>
        <th>Sample</th>
      </tr>
      {% for key, value in gts.items()|sort %}
      <tr>
        <td>{{ key }}</td>
        <td>{{ value["reference"] }}</td>
        <td>{{ value["sample"] }}</td>
      </tr>
      {% endfor %}
    </table>
</div>
<div id="qc">
    <div class="headline qchead">
        <h1>Quality Control</h1>
    </div>
    <p class="qcexplain explainer">The following quality control measurements were taken on the input sample during software processing.</p>
    <table class="qctable">
        <tr>
            <td>Passing reads</td>
            <td>{{ qc["passing reads"] }}</td>
        </tr>
        <tr>
            <td>Mean coverage</td>
            <td>{{ qc["Mean coverage"] }}</td>
        </tr>
        <tr>
            <td>>1X</td>
            <td>{{ qc["1x"] }}</td>
        </tr>
        <tr>
            <td>>5X</td>
            <td>{{ qc["5x"] }}</td>
        </tr>
        <tr>
            <td>>10X</td>
            <td>{{ qc["10x"] }}</td>
        </tr>
        <tr>
            <td>>100X</td>
            <td>{{ qc["100x"] }}</td>
        </tr>
        <tr>
            <td>Quality control</td>
            <td>{{ summary["qc"] }}</td>
        </tr>
    </table>
    <p class="qcpost explainer">The 1X, 5X, 10X, and 100X coverage measurements indicate the percentage of the reference genome targets covered at greater than or equal to the designated coverage number.</p>
</div>
<div id="legal">
    <div class="headline legalhead">
        <h1>Legal Information</h1>
    </div>
    <div class="sectionsbody">
        <h2 class="section">Quality control</h2>
        <p class="section-explain">{{ qc_methods }}</p>
        <h2 class="section">Disclaimer</h2>
        <p class="section-explain">{{ disclaimer }}</p>
        <h2 class="section">Methodology</h2>
        <p class="section-explain">{{ method }}</p>
        <h2 class="section">Limitations</h2>
        <p class="section-explain">{{ limitations }}</p>
        <h2 class="section">Laboratory certification</h2>
        <p class="section-explain">{{ cert }}</p>
        <h2 class="section">Confidentiality notice</h2>
        <p class="section-explain">{{ confident }}</p>
        <p class="section-explain">
            <b>{{ recipient }}</b>
        </p>
        <p class="section-explain">Approved by {{ approved }}</p>
        </div>
</div>
<!-- scripts -->
<!-- <script src="https://kit.fontawesome.com/874349d7b1.js" crossorigin="anonymous"></script> -->
<script src="https://code.jquery.com/jquery-3.4.1.slim.min.js" integrity="sha384-J6qa4849blE2+poT4WnyKhv5vZF5SrPo0iEjwBvKU7imGFAV0wwj1yYfoRSJoZ+n" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js" integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js" integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
</body>
</html>